File: c++/2272.cpp
// Link: https://leetcode.com/problems/substring-with-largest-variance/description/
// Runtime: 221 ms
// Memory: 6.9 MB

class Solution 
{
public:
    int largestVariance(const std::string& s) 
    {
        std::unordered_map<char, int> counter;
        for (const auto& symbol : s)
            counter[symbol]++;
        
        int largestVariance = 0;
        for (const auto& [major, value] : counter)
        {
            for (const auto& [minor, anotherValue] : counter)
            {
                if (major != minor)
                {
                    int majorCount = 0, minorCount = 0, restMinor = anotherValue;
                    for (const auto& symbol : s)
                    {
                        if (symbol == major) majorCount++;
                        else if (symbol == minor) 
                        {
                            minorCount++;
                            restMinor--;
                        }

                        if (minorCount > 0) 
                            largestVariance = std::max(largestVariance, majorCount - minorCount);
                        
                        if ((majorCount < minorCount) && (restMinor > 0))
                        {
                            majorCount = 0;
                            minorCount = 0;
                        }
                    }
                }
            }
        }

        return largestVariance;
    }
};

File: python/2272.py
# Link: https://leetcode.com/problems/substring-with-largest-variance/description/
# Runtime: 2907 ms
# Memory: 16.3 MB


class Solution:
    def largestVariance(self, s: str) -> int:
        counter = Counter(s)
        largest_variance = 0
        for major, minor in itertools.permutations(counter, 2):
            major_count, minor_count, rest_minor = 0, 0, counter[minor]
            for symbol in s:
                if (symbol == major) or (symbol == minor):
                    if major == symbol:
                        major_count += 1
                    elif minor == symbol:
                        minor_count += 1
                        rest_minor -= 1

                    if minor_count > 0:
                        largest_variance = max(
                            largest_variance, major_count - minor_count
                        )

                    if (major_count < minor_count) and (rest_minor > 0):
                        major_count, minor_count = 0, 0

        return largest_variance
